JustSaying/AwsTools/MessageHandling/MessageBatchInflightTracker.cs

```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.SQS.Model;
+
+namespace JustSaying.AwsTools.MessageHandling
+{
+    /// <summary>
+    /// Keeps track of a batch of messages that are inflight, as messages are signalled complete they are removed.
+    /// </summary>
+    public sealed class MessageBatchInflightTracker
+    {
+        readonly ConcurrentDictionary<string, Message> _messages;
+
+        public MessageBatchInflightTracker(IEnumerable<Message> messages)
+        {
+            _messages = new ConcurrentDictionary<string, Message>(
+                messages.Select(m => new KeyValuePair<string, Message>(m.MessageId, m)), StringComparer.Ordinal);
+        }
+
+        public void Complete(Message message) => _messages.TryRemove(message.MessageId, out _);
```

Comment:
It logically can't as far as I can see, though it might be worth asserting it with some logging.
the concurrent dictionary is to allow multiple threads to mutate the collection, but we have no risk of removing the same item concurrently or more than once.